### PR TITLE
Add parameterizable strategies

### PR DESCRIPTION
**Pull request by merlinND**
In this Pull Request, I add a common generic interface for parameterizable trading strategies.
Trading strategies declare their required and optional parameters. Each `Parameter` type defines its range of legal values.
This way, we prepare support for automatic strategy testing.

### CODE REVIEW

**Comment by merlinND** [comment]
Waiting on #27 before merging.

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  type Key = String
+  val parameters = params.toMap
+  
+  /**
+   * Similar to what a map's `get` would do.
+   * We also perform type checking.
+   */
+  def get[T : ClassTag](key: Key): Option[T] = {
+    parameters.get(key) match {
+      case Some(p) => p.get() match {
```
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: Do I understand currently that first you query by the name, and later you receive (concrete) parameter that actually has the name included too, so you have to do the second `.get()` to get the value?

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+ * @param T Parameter value type
+ */
+abstract class Parameter[T](val name: String) {
+  /**
+   * At construction, ensure that the given value is legal for this parameter.
+   */
+  assert(isValid, "Illegal value " + get() + " for strategy parameter " + companion.name)
+  
+  /** Retrieve the value for this parameter */
+  def get(): T
```
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: I would rename it to value actually.

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+  def getOrDefault[T : ClassTag](key: Key, parameterType: ParameterTrait[T]): T =
+    getOrElse(key, parameterType.defaultValue)
+  
+  override def toString: String = {
+    val strings = (for {
+      p <- parameters
+      key = p._1
+      paramType = p._2.toString()
+      value = p._2.get().toString()
+    } yield key + " (type " + paramType + ") = " + value).toList
```
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: Do you need toList at the end?

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  /**
+   * @return Whether the value is suitable for this parameter
+   */
+  def isValid(value: T): Boolean
+
+  /**
+   * Generate the range of valid values for this parameter.
+   * TODO: How to handle infinite ranges? Use streams?
+   */
+  def validValues: Iterable[T]
```
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala:147]: I would rather have the set of candidate values for each parameter (as opposed to all the valid values).

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala (hunk 3 above)
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: And also, doesn't `p._2.toString()` add the type and value too? Add a small test for that.

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	def get(): (Currency, Currency) = currencies
+}
+
+object CurrencyPairParameter extends ParameterTrait[(Currency, Currency)] {
+	private type CurrencyPair = Tuple2[Currency, Currency]
+	def getInstance(v: CurrencyPair) = new CurrencyPairParameter(v)
+  
+	/**
+	 * All currency pairs are acceptable as long as they're not twice the same.
+	 * @TODO: It would be great if we could enforce an order inside the pair, so that we avoid having to handle swapped currency pairs
```
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: I'm not sure about it, as someone may use this currency pair in a way that requires preserving the order. E.g. algorithm that always trades A->B, parametrized by (A,B)

File: ts/src/main/scala/ch/epfl/ts/traders/SimpleFXTrader.scala [hunk 6]
```diff
@@ -8,7 +8,7 @@ import ch.epfl.ts.data.{MarketAskOrder, MarketBidOrder, Quote}
 /* This simple trader will use two moving average and send order when this two MA cross each other.
  * @ Param the length of the two moving average period.
  */
-class SimpleFXTrader(val uid: Long, val shortPeriod: Int, val longPeriod : Int, val volume : Double) extends Component{
+class SimpleFXTrader(val uid: Long, val shortPeriod: Int, val longPeriod : Int, val volume : Double) extends Component {
```
Comment (sygi) [inline ts/src/main/scala/ch/epfl/ts/traders/SimpleFXTrader.scala]: :+1:

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    test(parameterTrait + " should give back the value it was instantiated with") {
+      for {
+        myValue <- validValues
+        parameter = parameterTrait.getInstance(myValue)        
+      } yield assert(parameter.get() == myValue, "Should give back value " + myValue)
+    }
+    
+    test(parameterTrait + " should reject invalid values") {
+      for {
+        myValue <- invalidValues
```
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: How are you going to define invalidValues? This sounds a bit crazy to me. I would just test some concrete examples (e.g. parameter is a numerical value => max(validValues) + 1000 is an invalid value).

Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: The other way may be to define a generator that would first generate an instance t of T and then yield if iff not isValidValue(t). Although I'm not sure how good the scala basic type generators are (e.g. if they try taking big values).

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  test("Should allow to add several parameters at a time") {
+    val myParameters = new StrategyParameters(
+      "tradedCurrencies" -> CurrencyPairParameter(currencies1),
+      "someCoefficient" -> CoefficientParameter(coefficient1),
+      "someOtherParameter" -> CurrencyPairParameter(currencies2)
+    )
+  }
+  
+  test("Should fail at instantiation with illegal parameters") {
+	  val attempt = Try(new StrategyParameters("someCoefficient" -> CoefficientParameter(coefficient2)))
```
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: rename `coefficient2` to `negativeCoefficient` or in other way mark that this is an illegal value. You may also consider moving it a bit closer to the test (or defining it in a test).

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    assert(attempt.isFailure, "Should fail to instantiate a coefficient with value " + coefficient2)
+  }
+  
+  test("Should hold the parameters and yield back their values") {
+	  val myParameters = new StrategyParameters(
```
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: indent

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      "tradedCurrencies" -> CurrencyPairParameter(currencies1),
+      "someCoefficient" -> CoefficientParameter(coefficient1)
+    )
+    
+    assert(myParameters.get[CurrencyPair]("tradedCurrencies") == Some(currencies1))
+    assert(myParameters.get[Double]("someCoefficient") == Some(coefficient1))
+  }
+  
+  test("Should not yield a value if it doesn't have the expected type") {
```
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala:169]: expected type -> corresponding key/value pair OR given value OR sth without the word type

Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala:169]: Ah, ok, now I get it.

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    val myParameters = new StrategyParameters("tradedCurrencies" -> CoefficientParameter(coefficient1))
+    
+    val got = myParameters.get[CurrencyPair]("tradedCurrencies")
```
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: You can also get the example `myParameters.get[CoefficientParameter]("unknown_key")` - it should behave the same, as far as I understand.

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala (hunk 7 above)
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: Ok, now I get it, removing signs of my ignorance.

File: ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+import ch.epfl.ts.data.CoefficientParameter
+import ch.epfl.ts.data.Parameter
+import ch.epfl.ts.data.ParameterTrait
+
+@RunWith(classOf[JUnitRunner])
+class StrategyParametersTests extends FunSuite {
+  /**
+   * Common test values
+   */
+	val currencies1 = (Currency.EUR, Currency.CHF)
```
Comment (sygi) [inline ts/src/test/scala/ch/epfl/ts/test/data/StrategyParameters.scala]: indent

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala (hunk 4 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala:147]: What is the difference between 'candidate' and 'valid' values?

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala (hunk 5 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: Oh, so all strategies should be robust to a currency swap in their parameters. That makes sense.

File: ts/src/main/scala/ch/epfl/ts/traders/SimpleFXTrader.scala (hunk 6 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/traders/SimpleFXTrader.scala]: :sparkles:

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala (hunk 1 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: The idea is:
The strategy demands a parameter under name with a certain type (key -> parameter). In this function, `T` gives the type of the value held by the parameter, `key` gives the name (e.g. `"tradingVolume"` or such).
Indeed, the second `get()` retrieves the actual value which was wrapped into the `Parameter` instance.

File: ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala (hunk 3 above)
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/data/StrategyParameters.scala]: You were right, thanks!

**Comment by liufengyun** [comment]
I see this is a cool feature for some scenarios, but it also complicates the usage and extension of TradingSimulation for basic scenarios.

Is it possible to keep the original traders unchanged for basic scenarios -- which I think it's the most used cases -- and use other approaches(decoration?) to support automatic strategy test?

**Comment by merlinND** [comment]
I understand your concern. If you have a look at https://github.com/merlinND/TradingSimulation/pull/31/files#diff-9cd994c5aec5b91dafdcbeeb91e1810bR50, you can see that the change to example code is quite minimal.
Moreover, if we want to support automatic testing, we need strategy designers to declare which parameters (in which valid range) do their strategy requires.
Therefore, supporting both interfaces would not really reduce the code that the designer has to write, and could be confusing. I think that enforcing a single interface is clearer, especially if it is well-described and justified in the doc.

**Comment by liufengyun** [comment]
Thanks for explanation. From the perspective of end users, my feeling is that though a helmet is helpful for riding bikes, it doesn't justify to force a user to wear helmet even in walking. It would be better it's an optional addon instead of a must. However, as you mentioned it's a tolerable overhead and there's no better alternatives, I accept the best solution we have.

**Comment by merlinND** [comment]
I love the analogy!
If we come across an elegant way to support both interfaces, I'm totally open to considering it.

**Comment by merlinND** [comment]
Auto :shipit:

**Pull request merged**